Control Themes

grokys opened the pull request:
## Background

The current biggest problem with our styling system is that there is no way to override an existing style without previously applied styles "leaking" through. This is especially important in the context of theming controls: if one has a globally-applied theme, then to re-theme a particular control or set of controls, one must reset every value set in the globally-applied theme. This can be impossible to do if you don't know what the globally-applied theme _is_ and also makes #7679 infeasible as the only way to reset these values is to use local values. This subject has been discussed in various threads:

- #2765
- #2769
- #5027
- #7120
- #7378

The two main proposals are contained in #7120 and #7378. This PR implements something more similar to #7120: I will go into the discussion we had over the pros and cons of each of these approaches, and why we chose this solution, in a separate comment.

## This PR

This PR implements a system similar to the WPF/UWP styling system, in addition to our existing CSS-like styling system:

- A `Theme` property was added to the `Control` class which accepts an instance of a `ControlTheme` object. This property is analogous to the WPF/UWP `FrameworkElement.Style` property.
- `ControlTheme` is a type of style that is placed in a `ResourceDictionary` instead of a `Styles` collection, and has a `TargetType` property instead of a `Selector`. This class is analogous to the WPF/UWP `Style` class.
  - It also has a `BasedOn` property
- If `Control.Theme` is not set, then the framework will attempt to find a `ControlTheme` resource with a key matching the control's style key
- `Control.Theme` can be specified inline, e.g.

```
<Button>
  <Button.Theme>
    <ControlTheme>
      <!-- Setters and nested styles here -->
    </ControlTheme>
  </Button.Theme>
</Button>
```

- `Control.Theme` can be set via a `StaticResource`, e.g.

```
<Button Theme="{StaticResource MyButtonTheme}"/>
```

- A `ControlTheme` can contain nested styles which are permitted to select on the control that has the theme and its _direct templated children_
- Control themes are applied at a lower priority than CSS-like styles

## Naming

This system unfortunately diverges from WPF/UWP's naming:

- The `FrameworkElement.Style` property becomes `Control.Theme`
- The `Style` class becomes `ControlTheme`

Ideally we would have used the same naming as WPF/UWP, but the problem is that we already have a `Style` class, and it has a `Selector`. A `Selector` is not valid on a `ControlTheme`, and similarly `ControlTheme` has `TargetType` and `BasedOn` properties which are not valid on styles.

Our choices here were:

- Find some other name for our existing `Style` class. This would be a huge API breakage.
- Use a single `Style` class with `Selector`, `TargetType` and `BasedOn` properties, and choose which ones are relevant depending on where the `Style` appears (in a `Styles` collection or a resource dictionary). This would be confusing.
- Find a new naming for the new properties/classes. This is what we chose.

## Todo

- [x] Validate selectors in control themes
- [x] `BasedOn`

## For subsequent PRs:

- Default control themes (i.e. Generic.xaml equivalent)
- Port fluent and simple theme

grokys committed:
Refactored most of Style into StyleBase.

Ready for `ControlTheme` class, which is a style without a selector.

grokys committed:
Initial implementation of control themes.

grokys committed:
Support ControlTheme in XAML compiler.

grokys committed:
Prevent ControlTheme as a nested style.

grokys committed:
Move Theme to StyledElement.

The WPF equivalent (`Style`) is in `FrameworkElement` so it would make sense. Will also make stuff a lot easier and removes the need for an `IThemed` interface.

grokys committed:
Promote theme to LocalValue if applied from style.

grokys committed:
Move tests to correct place.

grokys committed:
Invalidate template control styles when Theme changes.

grokys committed:
Add accent button to control catalog.

grokys committed:
Fix nested :not selector.

grokys committed:
Add ControlTheme.BasedOn.

grokys committed:
Display control themes in devtools.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021021-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Can apply control theme to derived types.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021024-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo commented:
Overall, I guess I'm not a fan of introducing another class here. It would have been better for Style to do all of this IMO.

robloo commented:
Also, I wonder why there is aversion to using Key like other XAML frameworks? It makes sense to be able to reference a ControlTemplate by Key in both BasedOn and directly in controls to choose which ControlTheme they want to apply overriding anything else.

maxkatz6 commented:
> Also, I wonder why there is aversion to using Key like other XAML frameworks

It isn't? ControlTheme needs to be stored somewhere, and it usually would be in ResourceDictionaries, and accessed by the key. Same as in WPF, where [FrameworkElement.Style](https://docs.microsoft.com/en-us/dotnet/api/system.windows.frameworkelement.style?view=windowsdesktop-6.0) is not a key, but an actual style object.

> It would have been better for Style to do all of this IMO

Yeah, but what to do with Selector property? It's confusing to have TargetType and Selector property, where Selector wouldn't work. Also, BasedOn won't work on old styles.
Since old avalonia styles (CssStyle) and control themes (WpfStyle) are quite difference by the way how they used, isn't it better to distingue them by name? It would be possible to rename old `Style` to `StyleSheet` (or something) and add new `Style` in this PR, but that's a way bigger breaking change that we would want to ever have.

maxkatz6 commented:
We had two options for naming as well: `ControlTheme` (and `Control.Theme` prop) or `ControlStyle` (and `Control.Style` prop).
Second one is closer to WPF API, but we once again have confusion with old Styles. Technically any old style is already a control's style, and some developers might be confused how old styles are related to `Control.Style` (and they aren't).

So, in the end new `ControlTheme` class is the least confusing one. Even though steps away from WPF naming.
Also, it would work better with `Control.ThemeVariant` property (alternative of [FrameworkElement.RequestedTheme](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.frameworkelement.requestedtheme?view=winrt-22621)), as it basically is a color variant of control theme.

Any other options possibly?

grokys committed:
Added support for implicit themes.

If no `Theme` property is provided, try to look up a resource keyed with the control's `StyleKey`.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021092-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Additional validation for ControlTheme children.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021142-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Merge branch 'fixes/8372-clear-local-value' into feature/7120-control-themes

grokys committed:
Added a new failing test.

grokys committed:
Merge branch 'fixes/8372-clear-local-value' into feature/7120-control-themes

grokys committed:
Add test/fix for promoted themes.

A bit of an edge case here, but we should deal with it.

grokys committed:
Added support for Design.PreviewWith in resource dictionaries.

robloo commented:
> Yeah, but what to do with Selector property? It's confusing to have TargetType and Selector property, where Selector wouldn't work. Also, BasedOn won't work on old styles.
>Since old avalonia styles (CssStyle) and control themes (WpfStyle) are quite difference by the way how they used, isn't it better to distingue them by name? It would be possible to rename old Style to StyleSheet (or something) and add new Style in this PR, but that's a way bigger breaking change that we would want to ever have.

I'm not convinced Selector can't be used in both cases. A Selector with only a control type is equivalent to TargetType. "ControlTheme" would have a restriction where Selector would only support a single type. Property name and syntax would be exactly the same -- just a much more narrow restriction on usage. 

Now we could argue conceptually there are differences between what a Selector and TargetType does. However, I would say Selector is a superset of a concept that includes TargetType. So where TargetType is valid, Selector is valid as a more generalized alternative.

> Also, BasedOn won't work on old styles.

Pretty sure it could. BasedOn would just require the selector types to match and would copy whatever BasedOn is pointing to and then add on top of that. Keep in mind Key should be added in Styles themselves to support this.

>  Selector is not valid on a ControlTheme, and similarly ControlTheme has TargetType and BasedOn properties which are not valid on styles.

Fundamentally, I think the existing styling system can be extended to do everything you are trying to do without adding new types to convolute things. Having ControlTheme along with Style seems like an unnecessary complexity to me and I'm not sure how that is going to look in usage in the years to come. Maybe we all get used to it. But maybe we also should just keep 1 Style like WPF.

Another fundamental issue I see is with the name "ControlTheme". The word "Theme" does imply the UI characteristics/design of a control (which could be the entire template along with styles like in WPF). However, in UWP and even Avalonia currently, it seems "Theme" just means the collection of base colors -- NOT the control templates. Additionally, a theme usually applies to all controls at once, not a single control. If a new class is really needed (and I don't think it fundamentally is) I would suggest a new name.

Honestly just using a name like "ControlStyle" would solve a few different concerns at once. It is telling you it's a style specific to an entire control. Where Styles themselves are treated just as fragments. The name is also closer to WPF and doesn't use the term "theme" which to me at least is counterintuitive.

amwx commented:
> I'm not convinced Selector can't be used in both cases. A Selector with only a control type is equivalent to TargetType. "ControlTheme" would have a restriction where Selector would only support a single type. Property name and syntax would be exactly the same -- just a much more narrow restriction on usage.
> 
> Now we could argue conceptually there are differences between what a Selector and TargetType does. However, I would say Selector is a superset of a concept that includes TargetType. So where TargetType is valid, Selector is valid as a more generalized alternative.

Technically, you're probably right, but my initial thought on this is Selector is a more expensive lookup rather than just a simple Type. Plus you'd have to have added validation to make sure the Selector is valid for "ControlTheme" - just type, no Classes, Pseudoclasses, ":not()", ":is()", or "/template/". And since part of this is about performance, it makes sense to try to avoid expensive look ups.


> Another fundamental issue I see is with the name "ControlTheme". The word "Theme" does imply the UI characteristics/design of a control (which could be the entire template along with styles like in WPF). However, in UWP and even Avalonia currently, it seems "Theme" just means the collection of base colors -- NOT the control templates. Additionally, a theme usually applies to all controls at once, not a single control. If a new class is really needed (and I don't think it fundamentally is) I would suggest a new name.
> 
> Honestly just using a name like "ControlStyle" would solve a few different concerns at once. It is telling you it's a style specific to an entire control. Where Styles themselves are treated just as fragments. The name is also closer to WPF and doesn't use the term "theme" which to me at least is counterintuitive.

I kind of agree here. I think `ControlStyle` would be a better choice. Especially after the other discussion on terminology in which ever thread that was where there was a clear desire to avoid "theme" since it doesn't have a consistent meaning anywhere.

<br />
<br />
I do wonder how maintaining effectively two style systems is gonna work out in the long run. And at least for the adoption phase of this, having ControlTheme at a lower priority than the css-style ones is going to cause problems somewhere - but that ordering makes sense and there really isn't anything to be done about that.

robloo commented:
> but my initial thought on this is Selector is a more expensive lookup rather than just a simple Type. Plus you'd have to have added validation to make sure the Selector is valid for "ControlTheme" - just type, no Classes, Pseudoclasses, ":not()", ":is()", or "/template/". And since part of this is about performance, it makes sense to try to avoid expensive look ups.

I don't think so. Consider the two cases:
 1. If ControlTheme or ControlStyle exists, it can do it's own parsing separately from other selectors. It would be very easy to handle this as a special case. The only thing that would really net to be preserved is the Syntax all parsing logic could differ and would be just as fast as parsing TargetType.
 2. If we are able to use Style for everything this gets a little more complicated. However, I don't think performance would be much of a concern in this case either. The first step of the Selector parsing logic would need to be a check for a Type match -- and only a type match. If that fails, the full parsing would be done. That would give this a special-cased hot path that really shouldn't be poor performance either.

> I do wonder how maintaining effectively two style systems is gonna work out in the long run. 

This is a good way of looking at it. Avalonia already separated Resources from Styles and now there will be two completely different "types" of Styles. That doesn't quite seem like a good idea although I understand some of the reasoning.

Another idea is to make a global DefaultStyleManager where styles could register themselves (and if more than 1, the last style registered would be the default). This could also be managed from code by themes. This would remove having the "Theme" or "Style" property on each control which also doesn't seem right to me.

maxkatz6 commented:
#### About merging features of Style and ControlTheme or "two style systems"

> I'm not convinced Selector can't be used in both cases. A Selector with only a control type is equivalent to TargetType. 

It would require compiler magic to restrict misusage of Selector on ControlTheme/ControlStyle. And there won't be a good solution for styles created in C#. At best there could be different style classes (as it is in this PR), but just rename TargetType to Selector, but that's confusing.

> I would say Selector is a superset of a concept that includes TargetType

Exactly for that reason it's not the best idea to use **super**set in both cases. Unless we change Selector to be a **sub**set, but I can't imagine it well now.
And if we allow full featured Selector support for control themes (or styles), it won't work well for performance as well, as @amwx mentioned.

> Pretty sure it could. BasedOn would just require the selector types to match and would copy whatever BasedOn is pointing to and then add on top of that. Keep in mind Key should be added in Styles themselves to support this.

And what's point of **Key** outside of resource dictionary? Or do you mean to be able to put one style without selector but with key in the resource dictionary, and reference it with BasedOn from styles with selectors? It makes sense, though we would need to force TargetType once again. Anyway, it seems to be a reasonable proposal, but probably outside of scope of this one.

> I think the existing styling system can be extended to do everything
> I do wonder how maintaining effectively two style systems is gonna work out in the long run.

There is no "two style systems". Selectors, nested styles, setters - that's all the 90% of shared styling mechanism, which wasn't changed (expect additions like nested styles). The only difference is the way how style group is applied on the control - either with a selector or with key.

As you can see, this PR is quite small to be called a new styling system.

#### Naming

> The word "Theme" does imply the UI characteristics/design of a control (which could be the entire template along with styles like in WPF). However, in UWP and even Avalonia currently, it seems "Theme" just means the collection of base colors -- NOT the control templates

Naming is hard, indeed.
But I don't agree about Avalonia part. In Avalonia "theme" term is used only in context of FluentTheme or DefaultTheme, which are precisely packages with collection of control templates and resources. If you add "ThemeVariant" for dark/light/high-contrast it also makes sense.
At the same time, you are absolutely correct, `ControlStyle` also makes sense, because as you said "It is telling you it's a style specific to an entire control". Especially when the `ControlStyle` shares lots of functionality with a `Style`.
Personally, I don't have a favorite here. CC @grokys @Takoooooo

maxkatz6 commented:
I like how `FluentTheme/DefaultTheme -> ControlTheme -> Control.Theme + Control.ThemeVariant` work together, making 'Theme' naming finally reasonable.

But at the same time Style and ControlStyle naming is less confusing as these mechanisms share the same idea and 90% of the implementation, with a difference only in how it's applied on the control. And I don't see confusion with StyledElement.Style. Except "Control.Theme" for dark/light mode is out of place now, but that's how UWP did it as well, so...

grokys commented:
Yeah we've been over this quite thoroughly in chat, and though we're not 100% happy with this solution and naming, it was generally felt to be the best compromise between getting the feature added and not breaking everyone.

I think @maxkatz6 covered most of it but regarding the naming, there were two main possibilities we discussed:

- `ControlStyle` with a `Style` property: we were worried that this would be confused with the existing styling system
  -  "I see there's a `Style` property on my button, how do I assign my `<Style Selector="Button">` to it?"
  - "My control is being styled by something but its `Style` property is `null`, what's happening?"
- `ControlTheme` with a `Theme` property
  - Despite the overloading of the term "theme" we don't currently have any `Theme` type which would be easily confused with something you can assign to this property
  - It will usually be used to apply a "Theme" to a "Control", i.e. a Control Theme

Given that the alternatives we discussed and threw out were renaming our existing `Style` classes and breaking our API massively, I think the naming we have here is the best we came up with.

robloo commented:
> And what's point of Key outside of resource dictionary? Or do you mean to be able to put one style without selector but with key in the resource dictionary, and reference it with BasedOn from styles with selectors? It makes sense, though we would need to force TargetType once again. Anyway, it seems to be a reasonable proposal, but probably outside of scope of this one.

Yes, I mean Key is literally used for lookup in BasedOn or Replaces and whatever other functionality is needed. It would allow one Style to reference another. As you said, the types also need to match. Key itself is only a string property and can be added easily to existing styles and used in lookup algos. You don't need any dictionary support for this.

> But I don't agree about Avalonia part. In Avalonia "theme" term is used only in context of FluentTheme or DefaultTheme, which are precisely packages with collection of control templates and resources.

In this case specifically, theme is a term that implies many controls are included. ControlTheme only works for one control. So how are you setting a full theme to a single control? So Control.Theme accepting a ControlTheme is really only using a piece of the overall theme -- you aren't setting the full theme itself. Hopefully that makes a bit more sense. Theme implies more than 1 control.

---

There is always something better, the only constraints we really have are time and compatibility (which are not insignificant I know). I've noticed every dev here acknowledges that the design isn't perfect and there are aspects that they are not happy with. That doesn't quite seem like a finished design to me.

I have experience with API design and naming and I'm not confident where we are now is what should be considered final. As stated, ControlTheme and Style are 90% the same. As software developers shouldn't the goal be to get that to 100%? Especially when doing so more closely matches existing conventions? Having a single Style is good in terms of fundamental design, code de-duplication, and compatibility. My gut feeling is anything less than that isn't a good final design. I would set this as the design objective if I was lead on this.

With that design objective in mind I wrote the past discussion #7378 outlining how to update the styling system achieving this (there were a few mistakes but I don't think any fundamental roadblocks). The vast majority of heavy lifting was already done with nested styles. With that, Style itself can be the container. We only need to add a few properties to Style now to finish things up. And yes, this is another reason TargetType and Selector should be the same -- You should only have Selector on Style to avoid confusion, duplication and new concepts to Avalonia.

I think the primary goal of this styling system is to have more (full) control over a control's default style/template. So how does Avalonia know which Style to apply to a certain control today? Well it figures this out by just applying every style that matches Selector, in-order, by scope (AFAIK). This is also the way other XAML frameworks do it more-or-less with some magic locations like generic.xaml. So what exists today is implicit default styles. Now we need to make that explicit.

There are many ways to do that without adding a new Theme property to define it directly on the control as a ControlTheme. The best idea to me would be to add a single property to Style "IsDefaultStyle". That would tell Avalonia it needs to register this style as the default for the indicated type. If no explicit style is registered, Avalonia could still fallback to implicit lookup that it does currently. Full compatibility.

From there we only need to add Key support in Style collections. Then BasedOn starts working along with OverridesDefaultStyle and Replaces if needed -- both of which just give more control over priority in the registered default style manager. The only difficultly I see here is managing scope but that's going to have to be managed anyway.

So it seems to me this is all being overcomplicated. We pre-empt issues of compatibility, "theme" naming and even developer confusion by just designing this PR differently from the start and extending already existing Styles.

Some of this also seems like you are trying to drop in WPF thinking without modifying it to fit in with the existing Avalonia styling system. Avalonia has already decided to go the CSS route and that has a lot of benefits over WPF. Don't look back now, keep pushing into new territory. Don't simply design a parallel ControlTheme similar to WPF and put it alongside.

grokys commented:
Ok, I'm not sure what to say. I've given you the reasoning behind the choices we made here, but you're not happy with them, and I don't think I have the energy to continue with this. I'll hand this feature over to someone else.

robloo commented:
@grokys Yes, you've given some reasoning but I'm not sure my proposal is understood (probably a failure of communication on my part). We never got into the details and it seems we were just discussing superficial topics. It is deeper than simply naming and even putting a theme property in controls. It was a fundamental re-think of the direction you've taken.

I also wouldn't say I am unhappy. What you and this project has achieved is astounding to say the least. My comments are purely technical and appologize if they seemed harsh or ignoring your points. I was just discussing concepts and ideas that differ. In a debate it's common to be passionate about viewpoints as well. But I do hope it isn't taken personally and throwing up hands saying "I give up" is clearly not where I wanted this to go.

About this feature specifically, I am putting up some resistance because I think I see a better way. Of course not everyone is going to agree but I didn't realize the discussion itself was going to be an issue (if i had the time or saw the chat conversatuon you referenced i would have made my comments earlier too).

If you want to ignore me and go down the path you were going on by all means do so. You certainly can overrule anything and you and the other maintainers think you have a better design. I do think we should be able to discuss things until consensus or agree to disagree is established though. Especially for something as fundamental as a styling system we are going to be stuck with for some time. Feature debates are common but perhaps I don't have the standing with the team to carry this conversation.

Anyway, I don't want to discourage your work. ControlTheme will fill the gaps in functionality and be a workable solution. As much as I think I see a better way, if there is no one else who sees what I see I can accept that.

grokys committed:
Fix parent selectors with /template/ at end.

Previously the combinator state was lost when traversing nested selectors. To fix it, instead of running the parent selector in `NestingSelector.Evaluate`, return the parent selector with `MovePrevious`. This required `MovePrevious` to be aware of parent styles and because I had to change its signature, I also made it internal as it doesn't need to be a public API.

grokys committed:
Make MovePreviousOrParent internal as well.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021473-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Added failing test.

grokys committed:
Apply own control theme before templated parent's.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021477-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

maxkatz6 committed:
Merge branch 'master' into feature/7120-control-themes

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021535-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Remove debug/unused code.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021544-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Don't run an unnecessary batch update.

If there are no styles to detach, there's no reason to run a batch update.

grokys committed:
Revert "Make MovePreviousOrParent internal as well."

This reverts commit 5e1f28f2a9e3a4f8606bf8efb6aa393d12f62d45.

grokys committed:
Revert "Fix parent selectors with /template/ at end."

This reverts commit e50b416d5bf1529544830d5a291a1fb2f1b0da64 as it was not functioning correctly.

grokys committed:
Skip these failing tests for now.

I will work on a fix.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021634-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Merge branch 'master' into feature/7120-control-themes

grokys committed:
Merge branch 'master' into feature/7120-control-themes

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021690-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Disallow selectors with trailing /template/.

It's an invalid selector: what does `Button /template/` select?

grokys committed:
Added ItemContainerTheme property.

Otherwise there's no easy way to apply themes to item containers.

grokys committed:
Added some ControlTheme benchmarks.

grokys committed:
Merge branch 'master' into feature/7120-control-themes

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0022230-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

danwalmsley commented on src/Avalonia.Base/StyledElement.cs:
```diff
@@ -590,6 +637,30 @@ protected virtual void OnInitialized()
+                // set via a style, it will get cleared! To work around this, if the value was
+                // applied at less than local value priority then promote the value to local value
+                // priority until styling is re-applied.
+                if (change.Priority > BindingPriority.LocalValue)
```
comment says less than, but condition is greater than... just checking my understanding here assume its probably correct.

maxkatz6 replied:
This enum goes in opposite way. Lower priority values have higher integer value.

maxkatz6 committed:
Merge branch 'master' into feature/7120-control-themes

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0022305-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]
